### PR TITLE
Use `context_header` for tail call perf improvement

### PR DESCRIPTION
**Pull request by saxena-anurag**
Fixes part of #3645

## Description

This PR adds (an optional) support for new `context_header`, where an extension can opt in to allocate an array of `uint64_t` of size 8 in the context passed to the eBPF program. The extra header allocated by the extension is then used by the eBPF runtime to store the eBPF state, instead of relying on per-CPU or per-thread runtime state to store tail call information.

### Perf Numbers:
With these changes, there is around **22**% improvement in the tail call numbers.

#### Baseline:
```
C:\performance>release\bpf_performance_runner.exe -i tests.yml -e .sys -r -t "bpf_tail_call"
Timestamp,Test,Average Duration (ns),CPU 0 Duration (ns),CPU 1 Duration (ns),CPU 2 Duration (ns),CPU 3 Duration (ns),CPU 4 Duration (ns),CPU 5 Duration (ns),CPU 6 Duration (ns),CPU 7 Duration (ns)
2024-07-11T03:26:31-0800,bpf_tail_call,278,279,282,275,279,276,278,277,279
```

#### After:
```
C:\performance>release\bpf_performance_runner.exe -i tests.yml -e .sys -r -t "bpf_tail_call"
Timestamp,Test,Average Duration (ns),CPU 0 Duration (ns),CPU 1 Duration (ns),CPU 2 Duration (ns),CPU 3 Duration (ns),CPU 4 Duration (ns),CPU 5 Duration (ns),CPU 6 Duration (ns),CPU 7 Duration (ns)
2024-07-11T03:24:22-0800,bpf_tail_call,215,218,214,215,215,215,215,216,215
```

### Code Changes:
1. Add capabilities field in `ebpf_program_data_t` where the program information provider (i.e. extension) can declare whether it supports `context_header`.
2. Changes in tail call and program invocation path to use "fast path" when the program information provider does support this.
3. Changes in PROG_ARRAY_MAP validation to ensure that all programs added to prog array map either support or dont support `context_header`.
4. Updated `sample ext`, `sock_addr`, `sock_ops` providers to support `context_header`

## Testing

Existing CICD. Also added tests to validate the below cases:
1. extension restart
2. prog_array map update

## Documentation

Updated documentation for extension development.

## Installation

No

### CODE REVIEW

File: libs/runtime/ebpf_object.h [hunk 1]
```diff
@@ -120,6 +127,7 @@ extern "C"
     typedef void (*ebpf_zero_ref_count_t)(ebpf_core_object_t* object);
     typedef void (*ebpf_free_object_t)(ebpf_core_object_t* object);
     typedef const ebpf_program_type_t (*ebpf_object_get_program_type_t)(_In_ const ebpf_core_object_t* object);
+    typedef const bool (*ebpf_object_get_context_header_support_t)(_In_ const ebpf_core_object_t* object);
```
Comment (Alan-Jowett) [inline libs/runtime/ebpf_object.h:130]: Why is this needed in the base object? Is it used for more than just program objects?

File: docs/eBpfExtensions.md [hunk 2]
```diff
@@ -100,6 +100,39 @@ structure from provided data and context buffers.
 * `context_destroy`: Pointer to `ebpf_program_context_destroy_t` function that destroys a program type specific
 context structure and populates the returned data and context buffers.
 * `required_irql`: IRQL at which the eBPF program is invoked by bpf_prog_test_run_opts.
+* `capabilities`: 32-bit integer describing the optional capabilities / features supported by the extension.
+
+**Capabilities**
+
+`PROGRAM_DATA_CAPABILITY_CONTEXT_HEADER`:
+
+Flag indicating that extension supports adding a context header at the start of each context passed to the eBPF program.
```
Comment (Alan-Jowett) [inline docs/eBpfExtensions.md:109]: Maybe make it clear that the pointer passed to the BPF program is after the context header?

File: libs/execution_context/ebpf_maps.c [hunk 3]
```diff
@@ -29,6 +29,7 @@ typedef struct _ebpf_core_object_map
     ebpf_lock_t lock;
     ebpf_map_definition_in_memory_t inner_template_map_definition;
     bool is_program_type_set;
+    bool supports_context_header;
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_maps.c:32]: Why are these on the base object and not the prog_array one?

**Review by Alan-Jowett — APPROVED** [review]
Maybe consider simplifying the object and map headers, but not blocked on this.

File: libs/execution_context/ebpf_maps.c (hunk 3 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:32]: Actually, scratch that. This is only being set for prog_array, hash_of_map, and array_of_map, so that's probably fine.

File: libs/runtime/ebpf_object.h (hunk 1 above)
Comment (saxena-anurag) [reply libs/runtime/ebpf_object.h:130]: As discussed, keeping this in base object, and also exposed a `context_header_support` for map objects laos.

File: docs/eBpfExtensions.md (hunk 2 above)
Comment (saxena-anurag) [reply docs/eBpfExtensions.md:109]: Updated.

File: include/ebpf_program_types.h [hunk 4]
```diff
@@ -87,7 +89,7 @@ typedef struct _ebpf_program_data
     ebpf_program_context_create_t context_create;   ///< Pointer to context create function.
     ebpf_program_context_destroy_t context_destroy; ///< Pointer to context destroy function.
     uint8_t required_irql;                          ///< IRQL at which the program is invoked.
-    bool context_header;                            ///< Whether the program context supports a context header.
+    uint32_t capabilities;                          ///< Capabilities supported by the program information provider.
```
Comment (dthaler) [inline include/ebpf_program_types.h:92]: Use a bitfield rather than `#define`s, for compile type safety.  defines are much more bug prone.
It's fine to union it with `uint32_t capabilities` if there's a reason to do so but line 12 is not recommended.

File: libs/execution_context/ebpf_program.c [hunk 5]
```diff
@@ -25,6 +25,8 @@
 static size_t _ebpf_program_state_index = MAXUINT64;
 #define EBPF_MAX_HASH_SIZE 128
 
+#define CAPABILITY_CONTEXT_HEADER(capabilities) (capabilities & PROGRAM_DATA_CAPABILITY_CONTEXT_HEADER)
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:28]: This define should also go away and just use something like
capabilities.program_data_compatibility.context_header
which also supports intellisense.

File: docs/eBpfExtensions.md [hunk 6]
```diff
@@ -100,6 +100,42 @@ structure from provided data and context buffers.
...
+An extension can choose to opt-in to support context header at the start of each program context structure that is
+passed to the eBPF program. To support this feature, the extension can use the macro `EBPF_CONTEXT_HEADER` to include
+the context header at the start of the program context structure. Even when the context header is added, the pointer
+passed to the eBPF program is after the context header.
+
+*Example*
+
+Below is an example of sample extension where it is now including context header start the of the original context
```
Comment (dthaler) [inline docs/eBpfExtensions.md:117]: I can't parse "header start the of the"

File: libs/execution_context/ebpf_program.c [hunk 7]
```diff
@@ -439,6 +455,28 @@ _ebpf_program_type_specific_program_information_attach_provider(
         goto Done;
     }
 
+    // Check if the context header support has changed. This check and behavior is needed for the below reasons:
+    // After a program has been loaded and attached, i.e. the program information provider and hook info providers
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:459]:
```suggestion
    // After a program has been loaded and attached, i.e., the program information provider and hook info providers
```

File: libs/execution_context/ebpf_program.h [hunk 8]
```diff
@@ -414,6 +417,39 @@ extern "C"
     size_t
     ebpf_program_get_state_index();
 
+    /**
+     * @brief Get whether the program information provider supports the context header.
+     *
+     * @param[in] program Pointer to the program object.
+     *
+     * @return true when the program supports the context header.
+     * @return false when the program does not support the context header.
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.h:426]:
```suggestion
     * @retval true The program supports the context header.
     * @retval false The program does not support the context header.
```
`return` should only appear once, but `retval` can appear once per return value.

File: tests/end_to_end/end_to_end.cpp [hunk 9]
```diff
@@ -3170,9 +3173,68 @@ extension_reload_test(ebpf_execution_type_t execution_type)
 
         // Program should not run.
         uint32_t hook_result = MAXUINT32;
-        REQUIRE(hook.fire(&ctx, &hook_result) != EBPF_SUCCESS);
+        REQUIRE(hook.fire(ctx, &hook_result) != EBPF_SUCCESS);
         REQUIRE(hook_result != 42);
     }
+
+    // Reload the extension again with original data
```
Comment (dthaler) [inline tests/end_to_end/end_to_end.cpp:3180]:
```suggestion
    // Reload the extension again with original data.
```

**Review by Alan-Jowett — DISMISSED** [review]

File: include/ebpf_program_types.h (hunk 4 above)
Comment (saxena-anurag) [reply include/ebpf_program_types.h]: Updated to use bitfield.

File: tests/end_to_end/helpers.h [hunk 10]
```diff
@@ -673,14 +673,7 @@ static ebpf_program_data_t _ebpf_xdp_test_program_data = {
...
 static ebpf_program_data_t _ebpf_bind_program_data = {
-    EBPF_PROGRAM_DATA_HEADER,
-    &_ebpf_bind_program_info,
-    NULL,
-    NULL,
-    NULL,
-    NULL,
-    0,
-    PROGRAM_DATA_CAPABILITY_CONTEXT_HEADER};
+    EBPF_PROGRAM_DATA_HEADER, &_ebpf_bind_program_info, NULL, NULL, NULL, NULL, 0, {true}};
```
Comment (dthaler) [inline tests/end_to_end/helpers.h:676]: can you do 
`{.supports_context_header = true}`
here which makes it clear what the "true" means?  Same question in 907

Comment (saxena-anurag) [reply tests/end_to_end/helpers.h]: Fixed.

File: docs/eBpfExtensions.md (hunk 6 above)
Comment (saxena-anurag) [reply docs/eBpfExtensions.md]: Updated the text.

File: docs/eBpfExtensions.md [hunk 11]
```diff
@@ -100,6 +100,43 @@ structure from provided data and context buffers.
...
 * `required_irql`: IRQL at which the eBPF program is invoked by bpf_prog_test_run_opts.
+* `capabilities`: 32-bit integer describing the optional capabilities / features supported by the extension.
+   * `supports_context_header`: Flag indicating extension supports adding a context header at the start of each context passed to the eBPF program.
+
+**Capabilities**
+
+`supports_context_header`:
+
+Flag indicating that extension supports adding a context header at the start of each context passed to the eBPF program.
+An extension can choose to opt-in to support context header at the start of each program context structure that is
```
Comment (dthaler) [inline docs/eBpfExtensions.md:111]:
```suggestion
An extension can choose to opt in to support context header at the start of each program context structure that is
```

File: libs/execution_context/ebpf_program.c [hunk 12]
```diff
@@ -1419,14 +1464,22 @@ ebpf_program_load_code(
...
+ebpf_program_set_tail_call(_In_ const void* context, _In_ const ebpf_program_t* next_program)
 {
     // High volume call - Skip entry/exit logging.
     ebpf_result_t result;
     ebpf_execution_context_state_t* state = NULL;
-    result = ebpf_state_load(_ebpf_program_state_index, (uintptr_t*)&state);
-    if (result != EBPF_SUCCESS) {
-        return result;
+
+    // BPF_PROG_ARRAY_MAP validates that either all programs in the map either support context header,
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:1473]:
```suggestion
    // BPF_PROG_ARRAY_MAP validates that either all programs in the map either support a context header,
```

File: libs/execution_context/ebpf_program.c [hunk 13]
```diff
@@ -1419,14 +1464,22 @@ ebpf_program_load_code(
...
+    // or none of them do. So, if the next program supports context header, then the current program
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:1474]:
```suggestion
    // or none of them do. So, if the next program supports a context header, then the current program
```

File: libs/execution_context/ebpf_program.c [hunk 14]
```diff
@@ -1419,14 +1464,22 @@ ebpf_program_load_code(
...
+    // must also support context header.
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:1475]:
```suggestion
    // must also support a context header.
```
grammar suggestions

File: libs/execution_context/ebpf_program.h [hunk 15]
```diff
@@ -186,6 +186,7 @@ extern "C"
      * @brief Invoke an ebpf_program_t instance.
      *
      * @param[in] program Program to invoke.
+     * @param[in] use_context_header Whether to use program context header to store state information.
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.h:189]:
```suggestion
     * @param[in] use_context_header Whether to use a context header to store state information.
```

File: tests/end_to_end/end_to_end.cpp [hunk 16]
```diff
@@ -3065,7 +3067,8 @@ extension_reload_test(ebpf_execution_type_t execution_type)
     // Should fail.
     REQUIRE(
         ebpf_program_load(
-            execution_type == EBPF_EXECUTION_NATIVE ? "test_sample_ebpf_um.dll" : "test_sample_ebpf.o",
+            // execution_type == EBPF_EXECUTION_NATIVE ? "test_sample_ebpf_um.dll" : "test_sample_ebpf.o",
```
Comment (dthaler) [inline tests/end_to_end/end_to_end.cpp:3070]: Shouldn't have commented out code,

Comment (saxena-anurag) [reply tests/end_to_end/end_to_end.cpp]: Removed the commented line

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]
